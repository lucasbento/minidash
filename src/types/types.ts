import { grid } from "blessed-contrib";

export interface CLIFlags {
  server: string;
  dashboardOnly: boolean;
  healthCheckOnly: boolean;
  mock: boolean;
}

export enum ResponseType {
  SUCCESSFUL = "SUCCESSFUL",
  FAILED = "FAILED",
  SERVER_UNREACHABLE = "SERVER_UNREACHABLE",
}

export interface LogLine {
  date: string;
  responseStatus: number | null;
  requestTime: number | null;
  responseType: ResponseType;
}

export interface LogsGroupedByHour {
  [H: string]: {
    numberOfSuccessResponses: number;
    numberOfFailedResponses: number;
  };
}

export interface ChartArgs {
  grid: grid;
  logsGroupedByHour: LogsGroupedByHour;
  logs: LogLine[];
}
