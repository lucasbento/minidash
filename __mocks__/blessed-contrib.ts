import contrib from "blessed-contrib";

export default {
  ...contrib,
  grid: jest.fn(),
};
