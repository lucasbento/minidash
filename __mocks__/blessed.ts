import blessed from "blessed";

export default {
  ...blessed,
  screen: jest.fn().mockReturnValue({
    key: jest.fn(),
    on: jest.fn(),
    render: jest.fn(),
  }),
};
