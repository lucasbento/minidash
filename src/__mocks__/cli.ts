const SERVER_URL = "google.com";

export { SERVER_URL };
