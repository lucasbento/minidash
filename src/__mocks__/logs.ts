import { LogLine, LogsGroupedByHour, ResponseType } from "../types/types";

const getLogLineWithDiffDate = (logLine: LogLine, date: string): LogLine => ({
  ...logLine,
  date,
});

const SingleLogLineMock: {
  [K in ResponseType]: LogLine;
} = {
  [ResponseType.SUCCESSFUL]: {
    date: "2020-12-31T11:00:43.284Z",
    responseStatus: 200,
    requestTime: 10,
    responseType: ResponseType.SUCCESSFUL,
  },
  [ResponseType.FAILED]: {
    date: "2020-12-31T11:00:43.284Z",
    responseStatus: 500,
    requestTime: 4,
    responseType: ResponseType.FAILED,
  },
  [ResponseType.SERVER_UNREACHABLE]: {
    date: "2020-12-31T11:00:43.284Z",
    responseStatus: null,
    requestTime: null,
    responseType: ResponseType.SERVER_UNREACHABLE,
  },
};

const MixedLogLinesMock = [
  // 12:00
  SingleLogLineMock[ResponseType.SUCCESSFUL],
  SingleLogLineMock[ResponseType.FAILED],
  SingleLogLineMock[ResponseType.FAILED],
  // 12:30
  getLogLineWithDiffDate(
    SingleLogLineMock[ResponseType.SERVER_UNREACHABLE],
    "2020-12-31T11:30:43.284Z"
  ),
  getLogLineWithDiffDate(
    SingleLogLineMock[ResponseType.SUCCESSFUL],
    "2020-12-31T11:30:43.284Z"
  ),
  getLogLineWithDiffDate(
    SingleLogLineMock[ResponseType.SUCCESSFUL],
    "2020-12-31T11:30:43.284Z"
  ),
  // 13:00
  getLogLineWithDiffDate(
    SingleLogLineMock[ResponseType.SERVER_UNREACHABLE],
    "2020-12-31T12:00:43.284Z"
  ),
  getLogLineWithDiffDate(
    SingleLogLineMock[ResponseType.SUCCESSFUL],
    "2020-12-31T12:00:43.284Z"
  ),
];

const LogsGroupedByHourMock: LogsGroupedByHour = {
  "12:00": {
    numberOfSuccessResponses: 1,
    numberOfFailedResponses: 2,
  },
  "12:30": {
    numberOfSuccessResponses: 2,
    numberOfFailedResponses: 1,
  },
  "13:00": {
    numberOfSuccessResponses: 1,
    numberOfFailedResponses: 1,
  },
};

export { SingleLogLineMock, MixedLogLinesMock, LogsGroupedByHourMock };
