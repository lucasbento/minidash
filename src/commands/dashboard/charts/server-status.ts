import blessed from "blessed";
import { ChartArgs, ResponseType } from "../../../types/types";

interface ServerStatus {
  [key: string]: {
    label: string;
    color: string;
  };
}

const SERVER_STATUS: ServerStatus = {
  [ResponseType.SUCCESSFUL]: {
    label: "UP!",
    color: "green",
  },
  [ResponseType.SERVER_UNREACHABLE]: {
    label: "UNREACHABLE!",
    color: "red",
  },
};

const getServerStatus = (logs: ChartArgs["logs"]) => {
  const lastLog = [...logs].splice(-1)[0];

  // If the last log is `FAILED`, check the last 5 logs for
  // response type as `SERVER_UNREACHABLE` or `FAILED`
  if (lastLog.responseType === ResponseType.FAILED) {
    const lastFiveLogs = [...logs].splice(-5);

    const allLogsAreFailedResponses = lastFiveLogs.every((log) =>
      [ResponseType.FAILED, ResponseType.SERVER_UNREACHABLE].includes(
        log.responseType
      )
    );

    if (allLogsAreFailedResponses) {
      return SERVER_STATUS[ResponseType.SERVER_UNREACHABLE];
    }

    // If they are not then just consider the server as "UP"

    return SERVER_STATUS[ResponseType.SUCCESSFUL];
  }

  // Server status will either be `Successful` or `Unreachable`
  return (
    SERVER_STATUS[lastLog.responseType] ||
    SERVER_STATUS[ResponseType.SERVER_UNREACHABLE]
  );
};

const createServerStatusChart = ({ grid, logs }: ChartArgs) => {
  const serverStatus = getServerStatus(logs);

  const serverStatusChart = grid.set(6.5, 8, 2, 4.15, blessed.box, {
    content: `\n{center}{bold}{${serverStatus.color}-fg}${serverStatus.label}{/${serverStatus.color}-fg}{/bold}{/center}`,
    tags: true,
    label: "Server status",
    border: {
      type: "line",
    },
    style: {
      fg: "blue",
      baseline: "white",
    },
  });

  return serverStatusChart;
};

export { createServerStatusChart };
