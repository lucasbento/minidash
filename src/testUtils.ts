import { mocked } from "ts-jest/utils";
import blessed from "blessed";
import contrib from "blessed-contrib";
import { MixedLogLinesMock, LogsGroupedByHourMock } from "./__mocks__/logs";

const mockedContrib = mocked(contrib, true);

const setData = jest.fn();
const gridSet = jest.fn().mockReturnValue({
  setData,
});

(mockedContrib.grid as jest.MockInstance<any, any>).mockImplementation(() => ({
  set: gridSet,
}));

const cleanUpMocks = () => {
  gridSet.mockClear();
  setData.mockClear();
};

const setupChartTest = () => {
  const screen = blessed.screen();
  const grid = new contrib.grid({ rows: 12, cols: 12, screen: screen });

  return {
    grid,
    logs: MixedLogLinesMock,
    logsGroupedByHour: LogsGroupedByHourMock,
    gridSet,
    setData,
    cleanUp: cleanUpMocks,
  };
};

export { setupChartTest };
