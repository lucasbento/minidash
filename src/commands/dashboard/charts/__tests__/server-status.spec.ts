import { setupChartTest } from "../../../../testUtils";
import { ResponseType } from "../../../../types/types";
import { SingleLogLineMock } from "../../../../__mocks__/logs";
import { createServerStatusChart } from "../server-status";

describe("createServerStatusChart", () => {
  const { grid, logs, logsGroupedByHour, gridSet, cleanUp } = setupChartTest();
  const chartArgs = {
    grid,
    logs,
    logsGroupedByHour,
  };

  afterEach(() => cleanUp());

  it("should create server status chart", () => {
    createServerStatusChart(chartArgs);

    expect(grid.set).toHaveBeenCalledTimes(1);
    expect(gridSet.mock.calls).toMatchSnapshot();
  });

  it("should create server status chart with status of `UNREACHABLE`", () => {
    createServerStatusChart({
      ...chartArgs,
      logs: [
        ...chartArgs.logs,
        SingleLogLineMock[ResponseType.SERVER_UNREACHABLE],
      ],
    });

    expect(grid.set).toHaveBeenCalledTimes(1);
    expect(gridSet.mock.calls).toMatchSnapshot();
  });

  it("should create server status chart with status of `UNREACHABLE` if there are 5 logs with FAILED `responseType`", () => {
    createServerStatusChart({
      ...chartArgs,
      logs: [
        ...chartArgs.logs,
        SingleLogLineMock[ResponseType.FAILED],
        SingleLogLineMock[ResponseType.FAILED],
        SingleLogLineMock[ResponseType.FAILED],
        SingleLogLineMock[ResponseType.FAILED],
        SingleLogLineMock[ResponseType.FAILED],
      ],
    });

    expect(grid.set).toHaveBeenCalledTimes(1);
    expect(gridSet.mock.calls).toMatchSnapshot();
  });

  it("should create server status chart with status of `UP` if there are less than 5 logs with FAILED `responseType`", () => {
    createServerStatusChart({
      ...chartArgs,
      logs: [
        ...chartArgs.logs,
        SingleLogLineMock[ResponseType.FAILED],
        SingleLogLineMock[ResponseType.FAILED],
        SingleLogLineMock[ResponseType.FAILED],
        SingleLogLineMock[ResponseType.FAILED],
      ],
    });

    expect(grid.set).toHaveBeenCalledTimes(1);
    expect(gridSet.mock.calls).toMatchSnapshot();
  });
});
