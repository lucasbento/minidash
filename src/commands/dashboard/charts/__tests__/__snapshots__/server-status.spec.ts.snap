// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createServerStatusChart should create server status chart 1`] = `
Array [
  Array [
    6,
    3.95,
    1.3,
    2.15,
    undefined,
    Object {
      "border": Object {
        "type": "line",
      },
      "content": "
{center}{bold}{green-fg}UP!{/green-fg}{/bold}{/center}",
      "label": "Server status",
      "style": Object {
        "baseline": "white",
        "fg": "blue",
      },
      "tags": true,
    },
  ],
]
`;

exports[`createServerStatusChart should create server status chart with status of \`UNREACHABLE\` 1`] = `
Array [
  Array [
    6,
    3.95,
    1.3,
    2.15,
    undefined,
    Object {
      "border": Object {
        "type": "line",
      },
      "content": "
{center}{bold}{red-fg}UNREACHABLE!{/red-fg}{/bold}{/center}",
      "label": "Server status",
      "style": Object {
        "baseline": "white",
        "fg": "blue",
      },
      "tags": true,
    },
  ],
]
`;

exports[`createServerStatusChart should create server status chart with status of \`UNREACHABLE\` if there are 5 logs with FAILED \`responseType\` 1`] = `
Array [
  Array [
    6,
    3.95,
    1.3,
    2.15,
    undefined,
    Object {
      "border": Object {
        "type": "line",
      },
      "content": "
{center}{bold}{red-fg}UNREACHABLE!{/red-fg}{/bold}{/center}",
      "label": "Server status",
      "style": Object {
        "baseline": "white",
        "fg": "blue",
      },
      "tags": true,
    },
  ],
]
`;

exports[`createServerStatusChart should create server status chart with status of \`UP\` if there are less than 5 logs with FAILED \`responseType\` 1`] = `
Array [
  Array [
    6,
    3.95,
    1.3,
    2.15,
    undefined,
    Object {
      "border": Object {
        "type": "line",
      },
      "content": "
{center}{bold}{green-fg}UP!{/green-fg}{/bold}{/center}",
      "label": "Server status",
      "style": Object {
        "baseline": "white",
        "fg": "blue",
      },
      "tags": true,
    },
  ],
]
`;
